extensions/amp-story/0.1/amp-story-auto-ads.js

```diff
@@ -29,42 +34,151 @@ export class AmpStoryAutoAds extends AMP.BaseElement {
     this.ampStory_ = null;
+
+    /** @private {?./navigation-state.NavigationState} */
+    this.navigationState_ = null;
+
+    /** @private {number} */
+    this.interactions_ = 0;
+
+    /** @private {!Array} */
+    this.adElements_ = [];
+
+    /** @private {number} */
+    this.adIdPointer_ = 0;
+
+    /** @private {?Array} */
+    this.pages_ = null;
   }
 
   /** @override */
   buildCallback() {
-    const ampStory = this.element.parentElement;
-    user().assert(ampStory.tagName === 'AMP-STORY',
+    const ampStoryElement = this.element.parentElement;
+    user().assert(ampStoryElement.tagName === 'AMP-STORY',
         `<${TAG}> should be child of <amp-story>`);
-    this.ampStory_ = ampStory;
-  }
+
+    ampStoryElement.getImpl().then(impl => {
+      this.ampStory_ = impl;
+      this.navigationState_ = this.ampStory_.getNavigationState();
+      this.navigationState_.observe(this.handleStateChange_.bind(this));
+    });
+
+    // TODO(ccordry): move this chunk to layoutCallback
+    // TODO(ccordry): need to fake distance from page to force load
+    for (let i = 0; i < 2; i++) {
+      const mockPage = this.makeMockPage();
+      this.adElements_.push(mockPage);
+      ampStoryElement.appendChild(mockPage);
+    }
+  };
+
 
   /** @override */
   isLayoutSupported() {
     return true;
   }
 
+
   /** @override */
-  layoutCallback() {
-    this.ampStory.appendChild(this.makeMockPage());
-  }
+  // layoutCallback() {}
+
 
   // temporary to be replaced with real page later
   makeMockPage() {
     const ampStoryAdPage = document.createElement('amp-story-page');
-    ampStoryAdPage.id = 'i-amphtml-ad-page-1';
+    const id = this.adElements_.length + 1;
+    ampStoryAdPage.id = `i-amphtml-ad-page-${id}`;
+    ampStoryAdPage.setAttribute('advertisement', '');
     ampStoryAdPage./*OK*/innerHTML = `
       <amp-story-grid-layer template="vertical">
-        <h1>First Ad Page</h1>
-        <p>This is the first ad shown in this story.</p>
+        <h1>Ad Page #${id}</h1>
+        <p>This is ad #${id} shown in this story.</p>
       </amp-story-grid-layer>
       `;
     return ampStoryAdPage;
   }
+
+
+  /**
+   * Get array containing all pages of associated story
+   * @private
+   */
+  getPages_() {
+    if (!this.pages_) {
+      this.pages_ = this.ampStory_.getPages();
+    }
+    return this.pages_;
+  }
+
+
+  /**
+   * Get array containing all pages of associated story
+   * @private
+   */
+  handleStateChange_(stateChangeEvent) {
+    switch (stateChangeEvent.type) {
+      case StateChangeType.ACTIVE_PAGE:
+        const {pageIndex, pageId} = stateChangeEvent.value;
+        this.handleActivePageChange_(
+            dev().assertNumber(pageIndex),
+            dev().assertString(pageId));
+        break;
+    }
+  }
+
+
+  /**
+   * @param {number} unusedPageIndex
+   * @param {string} unusedPageId
+   * @private
+   */
+  handleActivePageChange_(unusedPageIndex, unusedPageId) {
+    this.interactions_++;
```

Comment:
You'll want something more nuanced than this; right now, going back and forth between page 0 and page 1 multiple times will cause an ad to get slotted.  I guess you could measure how many unique page IDs have been seen since the last time an ad was slotted?  Without clearing the set of page IDs, but just clearing the count.  e.g.

```javascript
if (!this.seenPages_[pageId]) {
  this.seenPages_[pageId] = true;
  this.interactions_++;
}

// temp before config in passed in
if (this.interactions_ > MIN_INTERVAL) {
  this.placeNextAd_();
  this.interactions_ = 0;
}
```

Reply:
will refactor to set

Reply:
leaving this until we finalize our algorithm